### PR TITLE
Elasticsearch support for basic expressions on filter context

### PR DESCRIPTION
**Pull request by rafaelbey**
#### What type of PR is this?

Improvement

#### What does this PR do / why is it needed ?

This adds Elasticsearch support for basic expressions on filter context.

The expressions are handling by rewriting the functions, assigning the expression to a variable.  This allow the Pure platform and Allocation execution nodes to handle the expression evaluation.

This is limited to expressions that don't rely on the Elasticsearch fields, but other external parameters are accepted.

The main use case for this is to support Date operations like now, today, adjust, and conditional like if.

#### Which issue(s) this PR fixes:

Fixes #

### CODE REVIEW

File: legend-engine-core-language-pure/legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/handlers/Handlers.java [hunk 1]
```diff
@@ -472,7 +472,7 @@ public Handlers(PureModel pureModel)
...
                 h("meta::pure::functions::collection::take_T_MANY__Integer_1__T_MANY_", true, ps -> res(ps.get(0)._genericType(), "zeroMany"), ps -> true));
 
-        register("meta::pure::functions::lang::if_Boolean_1__Function_1__Function_1__T_m_", true, ps -> res(MostCommonType.mostCommon(Lists.fixedSize.of(funcReturnType(ps.get(1)), funcReturnType(ps.get(2))), this.pureModel), MostCommonMultiplicity.mostCommon(Lists.fixedSize.of(funcReturnMul(ps.get(1)), funcReturnMul(ps.get(2))), this.pureModel)));
+        register(h("meta::pure::functions::lang::if_Boolean_1__Function_1__Function_1__T_m_", true, ps -> res(MostCommonType.mostCommon(Lists.fixedSize.of(funcReturnType(ps.get(1)), funcReturnType(ps.get(2))), this.pureModel), MostCommonMultiplicity.mostCommon(Lists.fixedSize.of(funcReturnMul(ps.get(1)), funcReturnMul(ps.get(2))), this.pureModel)), ps -> Lists.immutable.of(MostCommonType.mostCommon(Lists.fixedSize.of(funcReturnType(ps.get(1)), funcReturnType(ps.get(2))), this.pureModel)), ps -> true));
```
Comment (rafaelbey) [inline legend-engine-core-language-pure/legend-engine-language-pure-compiler/src/main/java/org/finos/legend/engine/language/pure/compiler/toPureGraph/handlers/Handlers.java:475]: Handle type argument type on if function

File: legend-engine-pure-code/legend-engine-pure-code-compiled-core/src/main/resources/core/pure/executionPlan/executionPlan_generation.pure [hunk 2]
```diff
@@ -172,7 +172,7 @@ function <<access.private>> meta::pure::executionPlan::addFunctionParametersVali
    assert($collectionEnumParams->size() == 0, |'Collection of Enums is not supported as service parameter ' + $collectionEnumParams->map(p | $p.name)->makeString('[',', ',']'));
    let enumValidationContext = $planVarPlaceHolders->filter(p | $p.type->instanceOf(Enumeration))->map(e | ^meta::pure::executionPlan::EnumValidationContext(varName = $e.name, validEnumValues = $e.type->cast(@Enumeration<Any>)->enumValues()->map(e |$e->id())));
    let paramsSupportedForStreamInput = $routedFunction->findParamsSupportedForStreamInput(); 
-   if($f->instanceOf(LambdaFunction) && $planVarPlaceHolders->isNotEmpty(),| let functionParameters = $planVarPlaceHolders->map(p|^FunctionParameter(name=$p.name, supportsStream = $p.name->in($paramsSupportedForStreamInput), multiplicity=$p.multiplicity, type=$p.type));
+   if($planVarPlaceHolders->isNotEmpty(),| let functionParameters = $planVarPlaceHolders->map(p|^FunctionParameter(name=$p.name, supportsStream = $p.name->in($paramsSupportedForStreamInput), multiplicity=$p.multiplicity, type=$p.type));
```
Comment (rafaelbey) [inline legend-engine-pure-code/legend-engine-pure-code-compiled-core/src/main/resources/core/pure/executionPlan/executionPlan_generation.pure:175]: I don't understand why function parameter validation when the function is not a LambdaFunction.  This seems problematic, as there are valid cases to construct a plan from concreate functions, and these plans should validate the concreate function paramters.

File: legend-engine-pure-code/legend-engine-pure-code-compiled-core/src/main/resources/core/pure/protocol/vX_X_X/transfers/executionPlan.pure [hunk 3]
```diff
@@ -53,7 +53,7 @@ function meta::protocols::pure::vX_X_X::transformation::fromPureGraph::execution
             ^meta::protocols::pure::vX_X_X::metamodel::executionPlan::PureExpressionPlatformExecutionNode(
                _type='pureExp',
                resultType=$platform.resultType->transformResultType($mapping, $extensions),
-               pure = $platform.expression->meta::protocols::pure::vX_X_X::transformation::fromPureGraph::valueSpecification::transformValueSpecification([], newMap([])->cast(@Map<String, List<Any>>), $extensions)
+               pure = $platform.expression->meta::protocols::pure::vX_X_X::transformation::fromPureGraph::valueSpecification::transformValueSpecification([], newMap([]->cast(@Pair<String, List<Any>>)), $extensions)
```
Comment (rafaelbey) [inline legend-engine-pure-code/legend-engine-pure-code-compiled-core/src/main/resources/core/pure/protocol/vX_X_X/transfers/executionPlan.pure:56]: Legend Mixed support cannot infer the type arguments of the Map if cast is done after.  This cast of the pair parameter allow the proper inference of the Map type arguments.

File: legend-engine-xt-elasticsearch/legend-engine-xt-elasticsearch-V7-pure-metamodel/src/main/resources/core_elasticsearch_seven_metamodel/functions/pure_to_elasticsearch.pure [hunk 4]
```diff
@@ -945,19 +1024,51 @@ function meta::external::store::elasticsearch::v7::pureToEs::supportedFilterFunc
...
+
+  let platformNode = processValueSpecification($fe, ^PlatformPlanGenerationState(inScopeVars = $req.sq.inScopeVars, exeCtx = $req.exeCtx), $req.extensions, $req.debug)
+      ->toOne('Expression currently not supported: ' + $expression);
+
+  let allocNode = ^AllocationExecutionNode(resultType = $platformNode.resultType, varName = $varName, executionNodes = $platformNode);
+
+  pair(
+      list(^PlanVarPlaceHolder(type =  $platformNode.resultType.type, name = $varName, multiplicity = $fe.func->functionReturnMultiplicity())),
+      ^$req(extraNodes = $req.extraNodes->concatenate($allocNode))
+  );
```
Comment (rafaelbey) [inline legend-engine-xt-elasticsearch/legend-engine-xt-elasticsearch-V7-pure-metamodel/src/main/resources/core_elasticsearch_seven_metamodel/functions/pure_to_elasticsearch.pure:1071]: This is the main function to re-write the expression as a `let` and Elasticsearch will just reference the plan variable.

**Review by davidharte-gs — APPROVED** [review]

**Pull request merged**